### PR TITLE
Loki: Preserve pre-selected labels in the log context UI

### PR DESCRIPTION
**Pull request by ivanahuckova**
In this PR, we are introducing a new functionality to Loki log context UI. We are using local storage to save removed labels and added extracted labels and then use them in a new log context window. The main goal is to reduce repetition when users are going trough multiple log lines and they always have to select/remove same labels. 

Functionalities
- We've added button to reset query to initial query. It also deletes saved context labels from local storage
- We show standard Grafana notification when context is opened and previously selected labels were used
- If query with applied preserved labels will produce invalid stream selector without labels, we use original labels, we but don’t remove labels from local storage

https://github.com/grafana/grafana/assets/30407135/c727fcc1-8767-422a-94bf-abf6dbcaa292

### CODE REVIEW

File: public/app/plugins/datasource/loki/components/LokiContextUi.tsx [hunk 1]
```diff
@@ -143,8 +171,44 @@ export function LokiContextUi(props: LokiContextUiProps) {
...
+      // Do nothing when error occurs
+    } catch (e) {}
+
+    if (!preservedLabels) {
+      setContextFilters(initContextFilters);
+    } else {
+      // We need to update filters based on preserved labels
+      let arePreservedLabelsUsed = false;
+      const newContextFilters = initContextFilters.map((contextFilter) => {
+        if (preservedLabels && preservedLabels.removedLabels.includes(contextFilter.label)) {
```
Comment (svennergr) [inline public/app/plugins/datasource/loki/components/LokiContextUi.tsx:188]:
```suggestion
        if (preservedLabels.removedLabels.includes(contextFilter.label)) {
```

Should be covered by line 182 anyways.

File: public/app/plugins/datasource/loki/components/LokiContextUi.tsx [hunk 2]
```diff
@@ -143,8 +171,44 @@ export function LokiContextUi(props: LokiContextUiProps) {
...
+          arePreservedLabelsUsed = true;
+          return { ...contextFilter, enabled: false };
+        }
+
+        if (preservedLabels && preservedLabels.selectedExtractedLabels.includes(contextFilter.label)) {
```
Comment (svennergr) [inline public/app/plugins/datasource/loki/components/LokiContextUi.tsx:193]:
```suggestion
        if (preservedLabels.selectedExtractedLabels.includes(contextFilter.label)) {
```

Same as the above.

**Review by svennergr — APPROVED** [review]
Awesome, I really like that feature!

File: public/app/plugins/datasource/loki/components/LokiContextUi.tsx (hunk 1 above)
Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/components/LokiContextUi.tsx]: typescript is a little bit dumb here 

image

But I changed it to use `!` as we know at that point that we have preserved labels https://github.com/grafana/grafana/pull/68700/commits/6c8a00bed3ddcee16b901b07842eb1653d6c5b41

File: public/app/plugins/datasource/loki/components/LokiContextUi.tsx (hunk 2 above)
Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/components/LokiContextUi.tsx]: Same as above

**Pull request merged**